sample/redo/src/login.tsx

```diff
@@ -0,0 +1,85 @@
+import React, {useRef, useEffect, useCallback} from 'react';
+
+import {loginURL} from './settings';
+
+export type LoginResult = {
+  email: string;
+  userId: string;
+};
+
+async function loggedInUser(): Promise<LoginResult | undefined> {
+  const resJson = localStorage.loggedInUser;
+  if (resJson === undefined) {
+    return undefined;
+  }
+  return JSON.parse(resJson);
+}
+
+async function login(email: string) {
+  const resJson = localStorage.knownUsers;
+
+  const knownUsers = resJson !== undefined ? JSON.parse(resJson) : {};
+  let userId = knownUsers[email];
+  if (userId === undefined) {
+    userId = await remoteLogin(email);
+  }
+
+  knownUsers[email] = userId;
+
+  const user: LoginResult = {email, userId};
+
+  localStorage.loggedInUser = JSON.stringify(user);
```

Comment:
OMG, it only stores strings? I always thought localStorage could store any clonable object. WTAF.

Reply:
One would think so...